src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/PinnedGCHandle.T.cs

```diff
@@ -0,0 +1,148 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Represents an strongly-typed opaque, GC handle to a managed object.
+    /// A GC handle is used when an object reference must be reachable from
+    /// unmanaged memory.
+    /// The object is pinned at fixed location in GC heap and allows its
+    /// address to be taken.
+    /// </summary>
+    /// <remarks>
+    /// <see cref="PinnedGCHandle{T}"/> corresponds to Pinned roots.
+    /// For Normal, see <see cref="GCHandle{T}"/>.
+    /// For Weak and WeakTrackResurrection, see <see cref="WeakGCHandle{T}"/>.
+    /// </remarks>
+    /// <seealso cref="GCHandle" />
+    /// <typeparam name="T">The type of the object this <see cref="GCHandle{T}"/> tracks to.</typeparam>
+    public struct PinnedGCHandle<T> : IEquatable<PinnedGCHandle<T>>, IDisposable
+        where T : class?
+    {
+        // The actual integer handle value that the EE uses internally.
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Allocates a handle for the specified object.
+        /// </summary>
+        /// <param name="target">The object that uses the <see cref="GCHandle{T}"/>.</param>
+        public PinnedGCHandle(T target)
+        {
+            // Unlike GCHandle, pinning any object is allowed
+            _handle = GCHandle.InternalAlloc(target, GCHandleType.Pinned);
+        }
+
+        private PinnedGCHandle(IntPtr handle) => _handle = handle;
+
+        /// <summary>Determine whether this handle has been allocated or not.</summary>
+        public readonly bool IsAllocated => _handle != IntPtr.Zero;
+
+        /// <summary>Gets or sets the object this handle represents.</summary>
+        public readonly T Target
+        {
+            get
+            {
+                IntPtr handle = _handle;
+                GCHandle.CheckUninitialized(handle);
+                // Skip the type check to provide lowest overhead.
+                return Unsafe.As<T>(GCHandle.InternalGet(handle));
+            }
+            set
+            {
+                IntPtr handle = _handle;
+                GCHandle.CheckUninitialized(handle);
+                // Unlike GCHandle, pinning any object is allowed
+                GCHandle.InternalSet(handle, value);
+            }
+        }
+
+        /// <summary>
+        /// Retrieves the address of object data in a <see cref="PinnedGCHandle{T}"/>.
```

Comment:
This needs to include more details about what the object data is.

Also, it should include a note that this method should be only used for blittable types. The layout of non-blittable types is undefined for interop.